### PR TITLE
Update to guice 4.0

### PR DESCRIPTION
**Pull request by drcrallen**
- Mark a lot of `@Provides` methods as final since guice 4.0 disallows overriding them

### CODE REVIEW

File: extensions/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceModule.java [hunk 1]
```diff
@@ -71,7 +71,14 @@ public void setupModule(SetupContext setupContext)
...
   @LazySingleton
-  public Properties getProperties(
+  public final Properties getProperties(
+      @Json ObjectMapper mapper,
+      Properties systemProperties
+  ){
+    return getPropertiesInner(mapper, systemProperties);
+  }
+
+  protected Properties getPropertiesInner(
```
Comment (xvrl) [inline extensions/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceModule.java]: should we mark this method as to be used for testing only

**Comment by xvrl** [comment]
minor comment, but LGTM otherwise

File: extensions/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceModule.java (hunk 1 above)
Comment (drcrallen) [inline extensions/kafka-extraction-namespace/src/main/java/io/druid/server/namespace/KafkaExtractionNamespaceModule.java]: fixed

**Comment by nishantmonu51** [comment]
:+1:

**Comment by xvrl** [comment]
@drcrallen we should also update guice version in druid-api while we're at it for consistency, and update druid-api version

**Comment by drcrallen** [comment]
@xvrl : https://github.com/druid-io/druid-api/pull/51

**Comment by xvrl** [comment]
Sounds good, we can merge this as soon as druid-api with updated guice is released, and druid-api version has been updated in this PR.

**Comment by drcrallen** [comment]
build is expected to fail until the release for druid-api finishes.

**Comment by drcrallen** [comment]
@xvrl travis is passing here.

**Comment by himanshug** [comment]
@drcrallen  not certain but this PR seems to be causing following exception in hadoop batch ingestion (I'm using hadoop-2.6)
```
1) Error injecting method, java.lang.NoSuchMethodError: com.google.inject.spi.ProviderInstanceBinding.getUserSuppliedProvider()Ljavax/inject/Provider;
  at com.google.inject.multibindings.Multibinder$RealMultibinder.initialize(Multibinder.java:324)
  at io.druid.guice.PolyBind.optionBinder(PolyBind.java:113)
  while locating java.util.Set<java.util.Map$Entry<java.lang.String, com.google.inject.Provider<io.druid.metadata.MetadataSegmentPublisher>>>
  at com.google.inject.multibindings.MapBinder$RealMapBinder$RealProviderMapProvider.initialize(MapBinder.java:461)
  at io.druid.guice.PolyBind.optionBinder(PolyBind.java:113)
Caused by: java.lang.NoSuchMethodError: com.google.inject.spi.ProviderInstanceBinding.getUserSuppliedProvider()Ljavax/inject/Provider;
	at com.google.inject.multibindings.Indexer.visit(Indexer.java:139)
	at com.google.inject.multibindings.Indexer.visit(Indexer.java:43)
	at com.google.inject.internal.ProviderInstanceBindingImpl.acceptTargetVisitor(ProviderInstanceBindingImpl.java:62)
	at com.google.inject.multibindings.Multibinder$RealMultibinder.initialize(Multibinder.java:332)
	at com.google.inject.multibindings.Multibinder$RealMultibinder$$FastClassByGuice$$94efefc8.invoke(<generated>)
	at com.google.inject.internal.cglib.reflect.$FastMethod.invoke(FastMethod.java:53)
	at com.google.inject.internal.SingleMethodInjector$1.invoke(SingleMethodInjector.java:56)
	at com.google.inject.internal.SingleMethodInjector.inject(SingleMethodInjector.java:90)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:110)
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:75)
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:73)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.MembersInjectorImpl.injectAndNotify(MembersInjectorImpl.java:73)
	at com.google.inject.internal.Initializer$InjectableReference.get(Initializer.java:147)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.InjectorImpl$4$1.call(InjectorImpl.java:978)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.InjectorImpl$4.get(InjectorImpl.java:974)
	at com.google.inject.spi.ProviderLookup$1.get(ProviderLookup.java:89)
	at com.google.inject.spi.ProviderLookup$1.get(ProviderLookup.java:89)
	at com.google.inject.spi.ProviderLookup$1.get(ProviderLookup.java:89)
	at com.google.inject.multibindings.MapBinder$RealMapBinder$RealProviderMapProvider.initialize(MapBinder.java:469)
	at com.google.inject.multibindings.MapBinder$RealMapBinder$RealProviderMapProvider$$FastClassByGuice$$7fff4848.invoke(<generated>)
	at com.google.inject.internal.cglib.reflect.$FastMethod.invoke(FastMethod.java:53)
	at com.google.inject.internal.SingleMethodInjector$1.invoke(SingleMethodInjector.java:56)
	at com.google.inject.internal.SingleMethodInjector.inject(SingleMethodInjector.java:90)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:110)
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:75)
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:73)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1024)
	at com.google.inject.internal.MembersInjectorImpl.injectAndNotify(MembersInjectorImpl.java:73)
	at com.google.inject.internal.Initializer$InjectableReference.get(Initializer.java:147)
	at com.google.inject.internal.Initializer.injectAll(Initializer.java:92)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:173)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:109)
	at com.google.inject.Guice.createInjector(Guice.java:95)
	at com.google.inject.Guice.createInjector(Guice.java:72)
	at com.google.inject.Guice.createInjector(Guice.java:62)
	at io.druid.initialization.Initialization.makeInjectorWithModules(Initialization.java:412)
	at io.druid.indexer.HadoopDruidIndexerConfig.<clinit>(HadoopDruidIndexerConfig.java:93)
	at io.druid.indexer.DetermineHashedPartitionsJob$DetermineHashedPartitionsPartitioner.setConf(DetermineHashedPartitionsJob.java:393)
	at org.apache.hadoop.util.ReflectionUtils.setConf(ReflectionUtils.java:73)
	at org.apache.hadoop.util.ReflectionUtils.newInstance(ReflectionUtils.java:133)
	at org.apache.hadoop.mapred.MapTask$NewOutputCollector.<init>(MapTask.java:698)
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:767)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:163)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1628)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:158)
```

are you successfully being able to run hadoop ingestion after this update?

**Comment by drcrallen** [comment]
@himanshug Using my simple test of `index hadoop index-test.json` through my IntelliJ setup seems to work fine. Still investigating to see if the indexing service has something weird with this update

**Comment by drcrallen** [comment]
@himanshug Specifying `"hadoopCoordinates": "org.apache.hadoop:hadoop-client:2.6.0"` seems to be working locally. for the indexing service

**Comment by himanshug** [comment]
@drcrallen hmmm, not sure yet, but I have a hadoop-2.6 cluster deployed . ran druid processes with `hadoop classpath` output and hdfs-storage on classpath. doing the ingestion using HadoopIndexTask via overlord and it failed on me, removing the 2 commits from this PR solved the problem.

**Comment by xvrl** [comment]
@himanshug can you make sure nothing in your build or extensions is pulling in an older version of guice? This seems like somewhere on your classpath an older version of guice got pulled in.

**Comment by himanshug** [comment]
@xvrl that is true, "hadoop classpath" would include guice jar bundled with hadoop deployment, which will be guice-3.0 .
I think, probably I'm facing the issue because both guice-4.0 and guice-3.0 are ending up being on the classpath. In this case I'm using druid-services-0.8.1-SNAPSHOT-selfcontained.jar which will have guice-4.0 classes inside it.

**Comment by drcrallen** [comment]
@himanshug Showed up in our dogfood cluster. Reverting until IT can be fixed to catch issue.

**Comment by himanshug** [comment]
@drcrallen makes sense to revert till we figure out the root cause.

**Pull request merged**